Move HostAbortedException to Abstractions

eerhardt opened the pull request:
HostAbortedException was added directly to Microsoft.Extensions.Hosting, which depends on a lot of other libraries. This causes problems when using the HostFactoryResolver.Sources package because now consumers of HostFactoryResolver need to reference the whole Hosting package and all of its dependencies.

 Fix this by moving the new exception to the Hosting.Abstractions library, which has a minimal set of dependencies.

I couldn't add a dependency from `HostFactoryResolver.Sources` to `Hosting.Abstractions`. The packing infrastructure is getting in my way of doing that (cc @ViktorHofer). @dougbu - to consume `HostFactoryResolver.Sources`, you will need to manually add a reference to `Microsoft.Extensions.Hosting.Abstractions`.

eerhardt committed:
Move HostAbortedException to Abstractions

 HostAbortedException was added directly to Microsoft.Extensions.Hosting, which depends on a lot of other libraries. This causes problems when using the HostFactoryResolver.Sources package because now consumers of HostFactoryResolver need to reference the whole Hosting package and all of its dependencies.

 Fix this by moving the new exception to the Hosting.Abstractions library, which has a minimal set of dependencies.

stephentoub reviewed (approved):
Were there any tests that should be moved as well?

eerhardt commented:
> runtime (Build Browser wasm Linux Release LibraryTests) Failing after 68m

It's fitting that my PR fails for the flaky test that I'm trying to fix in [another PR](https://github.com/dotnet/runtime/pull/70185)

dougbu commented:
Please see https://github.com/dotnet/aspnetcore/pull/41999#issuecomment-1146141566. Short story is this won't work in aspnetcore and violates the purpose of the .Sources package.

> @dougbu - to consume `HostFactoryResolver.Sources`, you will need to manually add a reference to `Microsoft.Extensions.Hosting.Abstractions`.

Can't do this w/o limiting an "inside man" library to 7.0.0-preview6 or later web sites.

As I said in dotnet/aspnetcore#41999,
> May be better to either stick w/ only throwing `InvalidOperationException` but making `StopTheHostException` public.

deeprobin commented:
Imo we should introduce a test project for `Abstractions`-specific things.

But everything else looks good to me.

dougbu reviewed (changes requested):
Please find another way to make the `Exception` public

deeprobin commented:
> Please find another way to make the `Exception` public

why?

dougbu commented:
> > Please find another way to make the `Exception` public

> why?

Because any 7.0.x dependencies makes the .Sources package useless in https://github.com/dotnet/aspnetcore/blob/main/src/Tools/GetDocumentInsider/src/GetDocument.Insider.csproj and massively limits the applicable scope of the Microsoft.Extensions.ApiDescription.Client package. Please see https://github.com/dotnet/runtime/pull/70205#issuecomment-1146150610

eerhardt committed:
Remove direct dependency from HostFactoryResolver on Microsoft.Extensions.Hosting assemblies.

Load the HostAbortedException using Reflection instead. If it can't be found, throw a private exception with the same name.

dougbu commented on src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:
```diff
@@ -247,7 +247,7 @@ public object CreateHost()
                         _hostTcs.TrySetException(new InvalidOperationException("The entry point exited without ever building an IHost."));
                     }
-                    catch (TargetInvocationException tie) when (tie.InnerException is HostAbortedException)
+                    catch (TargetInvocationException tie) when (tie.InnerException?.GetType().Name == "HostAbortedException")
```
nit: Should this instead check the `Exception`'s full name❔

eerhardt replied:
The private exception purposefully has a different full name (since it is nested). So we would need to check 2 different "full names".

Even full names wouldn't prevent some other code running in a Host "Build()" sequence naming their exception `Microsoft.Extensions.Hosting.HostAbortedException`. So it isn't a guaranteed solution either.

If, for some reason, someone else names their exception `HostAbortedException`, and it isn't the one we threw below, the tooling will get `null` back from `HostFactoryResolver.ResolveServiceProviderFactory`, and it will handle it just like any other case where the IServiceProvider can't be found (usually displaying a message to the user saying something went wrong).

dougbu reviewed (approved):
Liking it 😺

eerhardt commented:
Merging to unblock aspnetcore taking a new runtime version. I can handle any follow up feedback in a separate PR.
